From 609e3c50cef1cfe09e9170ddc24d170ba2da246e Mon Sep 17 00:00:00 2001
From: GUO Zihua <guozihua@huawei.com>
Date: Tue, 19 Jul 2022 17:59:54 +0800
Subject: [PATCH] KEYS: Fix mistaken sizeof call in pgp_key_generate_id
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 609e3c50cef1cfe09e9170ddc24d170ba2da246e
Modified-by-SEL: No


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I5H61R
CVE: NA

--------------------------------

pgp_key_generate_id() is trying get the size of a flexible length
structure, however the sizeof() is called on the pointer itself.
Besides, considering it's trying to get the size of a flexible length
structure, use struct_size() instead.

Fixes: 4006f47d4e21 ("KEYS: PGP data parser")
Signed-off-by: GUO Zihua <guozihua@huawei.com>
Reviewed-by: Roberto Sassu <roberto.sassu@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 crypto/asymmetric_keys/pgp_public_key.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/crypto/asymmetric_keys/pgp_public_key.c b/crypto/asymmetric_keys/pgp_public_key.c
index 98b1707a0164..928029a13435 100644
--- a/crypto/asymmetric_keys/pgp_public_key.c
+++ b/crypto/asymmetric_keys/pgp_public_key.c
@@ -283,7 +283,8 @@ static struct asymmetric_key_ids *pgp_key_generate_id(
 		goto error;
 
 	kids->id[0] = kid;
-	kids->id[1] = kmemdup(kid, sizeof(kid) + fingerprint_len, GFP_KERNEL);
+	kids->id[1] = kmemdup(kid, struct_size(kid, data, fingerprint_len),
+			      GFP_KERNEL);
 	if (!kids->id[1])
 		goto error;
 
-- 
2.34.1

